### PR TITLE
Change format how settings represent lists / array

### PR DESCRIPTION
**Pull request by s1monw**
Today we represent each value of a list setting with it's own dedicated key that ends with the index of the value in the list. Aside of the obvious weirdness this has several issues especially if lists are massive since it causes massive runtime penalties when validating settings. Like a list of 100k words will literally cause a create index call to timeout and in-turn massive slowdown on all subsequent validations runs. 

This change moves away from the current internal representation towards a single decoded string representation. A list is encoded into a JSON list internally in a fully backwards compatible way. A list is then a single key value pair in settings and all prefix based settings are internally converted to the new representation in the settings builder. This change also forbids to add a settings that ends with a `.0` which was internally used to detect a list setting. Once this has been rolled out for an entire major version all the internal `.0` handling can be removed since all settings will be converted.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/settings/Settings.java [hunk 1]
```diff
@@ -739,6 +745,14 @@ public Builder put(Object... settings) {
          * @return The builder
          */
         public Builder put(String key, String value) {
+            if (key.endsWith(".0")) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/Settings.java]: This makes me uncomfortable.... I see that we used to do it as well. Can we get away with not doing it anymore now that we have `putArray`?

File: core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java [hunk 2]
```diff
@@ -407,21 +407,21 @@ public void testDiff() throws IOException {
         ClusterSettings settings = new ClusterSettings(Settings.EMPTY, new HashSet<>(Arrays.asList(fooBar, fooBarBaz, foorBarQuux,
             someGroup, someAffix)));
         Settings diff = settings.diff(Settings.builder().put("foo.bar", 5).build(), Settings.EMPTY);
-        assertEquals(4, diff.size()); // 4 since foo.bar.quux has 3 values essentially
+        assertEquals(2, diff.size()); // 4 since foo.bar.quux has 3 values essentially
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java]: No need for the comment anymore.

File: core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java [hunk 3]
```diff
@@ -407,21 +407,21 @@ public void testDiff() throws IOException {
...
         assertThat(diff.getAsInt("foo.bar.baz", null), equalTo(17));
         assertArrayEquals(diff.getAsArray("foo.bar.quux", null), new String[] {"d", "e", "f"});
 
         diff = settings.diff(
             Settings.builder().put("some.group.foo", 5).build(),
             Settings.builder().put("some.group.foobar", 17, "some.group.foo", 25).build());
-        assertEquals(6, diff.size()); // 6 since foo.bar.quux has 3 values essentially
+        assertEquals(4, diff.size()); // 6 since foo.bar.quux has 3 values essentially
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java]: Same. These comments were helpful but are not incorrect.

File: core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java [hunk 4]
```diff
@@ -407,21 +407,21 @@ public void testDiff() throws IOException {
...
-        assertEquals(4, diff.size()); // 4 since foo.bar.quux has 3 values essentially
+        assertEquals(2, diff.size()); // 4 since foo.bar.quux has 3 values essentially
         assertThat(diff.getAsInt("foo.bar.baz", null), equalTo(1));
         assertArrayEquals(diff.getAsArray("foo.bar.quux", null), new String[] {"a", "b", "c"});
 
         diff = settings.diff(
                 Settings.builder().put("foo.bar", 5).build(),
                 Settings.builder().put("foo.bar.baz", 17).putArray("foo.bar.quux", "d", "e", "f").build());
-        assertEquals(4, diff.size()); // 4 since foo.bar.quux has 3 values essentially
+        assertEquals(2, diff.size()); // 4 since foo.bar.quux has 3 values essentially
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java]: Here too.

File: core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java [hunk 5]
```diff
@@ -432,7 +432,7 @@ public void testDiff() throws IOException {
             Settings.builder().put("some.prefix.foo.somekey", 5).build(),
             Settings.builder().put("some.prefix.foobar.somekey", 17,
                 "some.prefix.foo.somekey", 18).build());
-        assertEquals(6, diff.size()); // 6 since foo.bar.quux has 3 values essentially
+        assertEquals(4, diff.size()); // 6 since foo.bar.quux has 3 values essentially
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java]: And here.

**Review by nik9000 — COMMENTED** [review]
I looked at it and thought about it some. I've left what I thought. I'll come back and have another look soon.

**Comment by rjernst** [comment]
When discussing this many weeks ago, I suggested having the value of the internal map be a discriminated union.  This could either simply be Object with the discrimination done by instanceof checks, or a simple POJO with a boolean+Object.  I'm apprehensive about this PR because it means every iteration of a list setting requires creation of new strings (by parsing the "list" string).  Why are these special marker strings to denote whether the value is a list any better than actually having a List and using instanceof? In either case, we have to ensure access to the internal map is removed, but the latter is much more transparent and easier to not mess up.

**Comment by s1monw** [comment]
I am closing this one since the encapsulation of Settings is so broken today I first have to clean up all the cruft around it to make it work. I think ultimately we can just have a `String,Object` map internally but before that we have to get rid of things like `getAsMap` on settings.

**Pull request closed without merging**